src/hooks/useOutputMaterial.ts

```diff
@@ -52,59 +87,74 @@ export function useOutputMaterial(sessionId: string, outputIdOrName: string, mat
 	 */
 	outputNode: ITreeNode | undefined
 } {
-
-	const materialPropertiesRef = useRef(materialProperties);
+	const materialPropertiesRef = useRef<IMaterialAbstractDataProperties | undefined>(undefined);
+	const materialTypeRef = useRef<MaterialType | undefined>(undefined);
 	
-	// callback which will be executed on output update, and once the node is available
-	const callback = useCallback( (node?: ITreeNode) => {
-		const materialProps = materialPropertiesRef.current;
-		if (!node)
-			return;
-
-		// get all geometry and materials below the node
-		const geometryData = getGeometryData(node);
-		const materials = getMaterials(geometryData);
+	// callback which will be executed on update of the output node
+	const callback = useCallback( (newNode?: ITreeNode, oldNode?: ITreeNode) => {
 	
-		if (materials.length > 0) {
-			// update existing materials
-			materials.forEach((m) => {
-				for ( const p in materialProps)
-					(<any>m)[p as keyof MaterialStandardData] =
-					materialProps[p as keyof IMaterialStandardDataProperties];
-				m.updateVersion();
+		// restore original materials
+		// TODO there seems to be a bug, oldNode.id should not be a new id
+		console.debug("newNode.id", newNode?.id, "oldNode?.id", oldNode?.id, originalMaterials);
+		if (oldNode && originalMaterials[oldNode.id]) {
+			const geometryData = getGeometryData(oldNode);
+		
+			geometryData.forEach(data => {
+				const originalMaterial = originalMaterials[oldNode.id][data.id];
+				if (originalMaterial) {
+					data.material = originalMaterial;
+					data.updateVersion();
+				}
 			});
+
+			delete originalMaterials[oldNode.id];
+			// TODO remove this
+			console.debug("restore oldNode", Object.keys(originalMaterials));
+
+			oldNode.updateVersion();
 		}
-		else
-		{
-			// no material found: define new material
-			const material = new MaterialStandardData(materialProps);
-			geometryData.forEach(data => data.material = material);
-		}
 
-		node.updateVersion();
+		// create and set the new material
+		if (newNode && materialPropertiesRef.current && materialTypeRef.current) {
+			const geometryData = getGeometryData(newNode);
+
+			// backup original materials
+			if (!originalMaterials[newNode.id]) {
+				originalMaterials[newNode.id] = {};
+				geometryData.forEach(data => {
+					originalMaterials[newNode.id][data.id] = data.material;
+				});
+			}
+
+			// TODO do we need to newly create the material on every update?
+			const newMaterial = createMaterial(materialPropertiesRef.current, materialTypeRef.current);
```

Comment:
@MajorMeerkatThe3rd say the output gets updated due to parameter changes several times: Do we need to create a new material for every update, or can we just assign the same? If so, I will change the call to create the material only once if the material properties change.